### PR TITLE
[MU4] Improve the dot adjustment algorithm

### PR DESCRIPTION
**Pull request by asattely**
Dots can now adjust for other dots in the chord:
image

Also, downstemmed voices bias downwards instead of upwards:
image

Resolves: https://github.com/musescore/MuseScore/issues/16210, too

### CODE REVIEW

**Review by oktophonie — APPROVED** [review]

File: src/engraving/libmscore/note.cpp [hunk 1]
```diff
@@ -2133,48 +2133,138 @@ void Note::setHeadHasParentheses(bool hasParentheses)
 
 void Note::setDotY(DirectionV pos)
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/note.cpp:2134]: This function looks now too overcomplicated. Too many nested conditions and loops. I would recommend moving nested code to separate private methods

File: src/engraving/libmscore/note.cpp [hunk 2]
```diff
@@ -2133,48 +2133,138 @@ void Note::setHeadHasParentheses(bool hasParentheses)
...
+        // construct combined chords using the notes from overlapping chords
+        if (!hasVoices) {
+            // only this voice, so topDownNotes is just the notes in the chord
+            for (Note* note : c->notes()) {
+                if (note->line() & 1) {
+                    int newOffset = 0;
+                    bool adjustDown = (c->voice() & 1) && !c->up();
+                    if (!anchoredDots.empty() && anchoredDots.back() == note->line()) {
+                        if (anchoredDots.size() >= 1 && anchoredDots[anchoredDots.size() - 2] == note->line() + adjustDown ? 2 : -2) {
+                            newOffset = adjustDown ? -2 : 2;
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/note.cpp:2179]: What does these numbers (-2 and 2) mean?

File: src/engraving/libmscore/note.cpp [hunk 3]
```diff
@@ -2133,48 +2133,138 @@ void Note::setHeadHasParentheses(bool hasParentheses)
...
         } else {
-            y = 0.5;
+            // Get a list of notes in this staff that adjust dots from top down,
+            // bottom up, and also start our locked-in dot list by adding all lines where dots are
+            // guaranteed
+            Measure* m = c->measure();
+            size_t firstVoice = c->track() - c->voice();
+            for (size_t i = firstVoice; i < firstVoice + VOICES; ++i) {
+                if (Chord* voiceChord = m->findChord(c->tick(), i)) {
+                    /*if (voiceChord->dots() == 0) {
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/note.cpp:2197]: Please remove this commented code

File: src/engraving/libmscore/note.cpp [hunk 4]
```diff
@@ -2133,48 +2133,138 @@ void Note::setHeadHasParentheses(bool hasParentheses)
...
+                        for (Note* note : voiceChord->notes()) {
+                            if (note->line() & 1) {
+                                anchoredDots.push_back(note->line());
+                            } else {
+                                topDownNotes.push_back(note);
+                            }
+                        }
+                    } else {
+                        for (Note* note : voiceChord->notes()) {
+                            if (note->line() & 1) {
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/note.cpp:2211]: Similar code (with & 1) is present in a lot of places. What does it mean?

File: src/engraving/libmscore/note.cpp [hunk 5]
```diff
@@ -2133,48 +2133,138 @@ void Note::setHeadHasParentheses(bool hasParentheses)
...
+        // our two lists now contain only notes that are on lines
+        std::sort(topDownNotes.begin(), topDownNotes.end(),
+                  [](Note* n1, Note* n2) { return n1->line() < n2->line(); });
+        std::sort(bottomUpNotes.begin(), bottomUpNotes.end(),
+                  [](Note* n1, Note* n2) { return n1->line() > n2->line(); });
+
+        bool finished = false;
+        for (Note* note : topDownNotes) {
+            int dotMove = -1;
+            if (std::find(anchoredDots.begin(), anchoredDots.end(), note->line() + dotMove) != anchoredDots.end()) {
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/note.cpp:2230]: you can use mu::contains instead

File: src/engraving/libmscore/note.cpp (hunk 4 above)
Comment (asattely) [reply src/engraving/libmscore/note.cpp:2211]: Odd/even check

File: src/engraving/libmscore/note.cpp (hunk 2 above)
Comment (asattely) [reply src/engraving/libmscore/note.cpp:2179]: Number of staff spaces or lines to adjust, where the top line is 0, then the top space is 1, etc etc

**Comment by asattely** [comment]
Implemented all of @RomanPudashkin's fixes and rebased to resolve the vtest name conflict!

File: src/engraving/libmscore/note.cpp (hunk 1 above)
Comment (asattely) [reply src/engraving/libmscore/note.cpp:2246]: Done!

File: src/engraving/libmscore/note.cpp (hunk 5 above)
Comment (asattely) [reply src/engraving/libmscore/note.cpp]: Done!

**Comment by RomanPudashkin** [comment]
@asattely please rebase it

**Comment by asattely** [comment]
rebased!

**Comment by Jojo-Schmitz** [comment]
Won't the unit tests need fixing?

**Comment by cbjeukendrup** [comment]
Looks like a rebase should be enough for that, just like in https://github.com/musescore/MuseScore/pull/15383.

**Review by abariska — APPROVED** [review]
Performance is tested

**Pull request merged**